From 32621f6fe6cad21c6a56d371cea9eedc00853fc5 Mon Sep 17 00:00:00 2001
From: Huang Chaofan <chaofan.huang@nxp.com>
Date: Tue, 3 Apr 2018 16:11:46 +0800
Subject: [PATCH 3564/5242] MLK-17974[IMX8QXP B0] Add judgement for
 find_buffer_id to avoid read NULL buffer

commit  d80d4225a7a80d3707d1c5d15cb21172083739a4 from
https://source.codeaurora.org/external/imx/linux-imx.git

Add judgement for find_buffer_id to avoid read NULL buffer, it is a
simple workaround and this case should not happen

Signed-off-by: Huang Chaofan <chaofan.huang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mxc/vpu-decoder-b0/vpu_b0.c |   12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/drivers/mxc/vpu-decoder-b0/vpu_b0.c b/drivers/mxc/vpu-decoder-b0/vpu_b0.c
index 99784db..cd01b20 100755
--- a/drivers/mxc/vpu-decoder-b0/vpu_b0.c
+++ b/drivers/mxc/vpu-decoder-b0/vpu_b0.c
@@ -175,9 +175,12 @@ static int find_buffer_id(struct vpu_ctx *ctx, u_int32 addr)
 		p_data_req = &ctx->q_data[V4L2_DST].vb2_reqs[i];
 		if (p_data_req->vb2_buf != NULL) {
 			pphy_address = (u_int32 *)vb2_plane_cookie(p_data_req->vb2_buf, 0);
-			LumaAddr = *pphy_address;
-			if (LumaAddr == addr - ctx->dev->cm_offset)
-				return i;
+			if (pphy_address != NULL) {
+				LumaAddr = *pphy_address;
+				if (LumaAddr == addr - ctx->dev->cm_offset)
+					return i;
+			} else
+				vpu_dbg(LVL_ERR, "error: %s() buffer (%d) is NULL\n", __func__, i);
 		}
 	}
 
@@ -1283,6 +1286,9 @@ static void report_buffer_done(struct vpu_ctx *ctx, void *frame_info)
 
 	buffer_id = find_buffer_id(ctx, FrameInfo[1]);
 
+	if (buffer_id == -1)
+		return;
+
 	if (buffer_id != fs_id)
 		vpu_dbg(LVL_ERR, "error: find buffer_id(%d) and firmware return id(%d) doesn't match\n",
 				buffer_id, fs_id);
-- 
1.7.9.5

